### PR TITLE
[#44] Add support for @ElementCollection

### PR DESCRIPTION
**Pull request by blafond**
Added support for List<String>, Set<String> and List<{Embeddable Entity}>

Tests include:
-  insert, remove, replace, clear elements
- duplicate and null elements.
- clearing of generated collection tables on delete of primary table

Will be logging issues to address support for
- @OrderColumn
- Map<> types ElementCollections
- Embeddable entity with a collection of elements
- additional tests for each

### CODE REVIEW

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+								locOffset = writeIndex( st, collection.getIndex( entry, i, this ), locOffset, session );
+							}
+							locOffset = writeElement( st, collection.getElement( entry ), locOffset, session );
+
+							collection.afterRowInsert( this, entry, i );
+						}
+				);
+				loop = loop.thenCompose( total -> reactiveConnection
+						.update( getSQLInsertRowString(), paramValues )
+						.thenApply( insertCount -> total + insertCount ) );
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:297]: I guess we had better trampoline this loop.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/collection/impl/ReactiveBatchingCollectionInitializerBuilder.java [hunk 2]
```diff
@@ -24,7 +24,7 @@
 public abstract class ReactiveBatchingCollectionInitializerBuilder {
 
 	public static ReactiveBatchingCollectionInitializerBuilder getBuilder(SessionFactoryImplementor factory) {
-		switch ( factory.getSettings().getBatchFetchStyle() ) {
+		switch ( factory.getSessionFactoryOptions().getBatchFetchStyle() ) {
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/collection/impl/ReactiveBatchingCollectionInitializerBuilder.java:27]: What's the reason for this change?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import org.hibernate.event.spi.EventType;
+import org.hibernate.event.spi.PostCollectionRecreateEvent;
+import org.hibernate.event.spi.PostCollectionRecreateEventListener;
+import org.hibernate.event.spi.PreCollectionRecreateEvent;
+import org.hibernate.event.spi.PreCollectionRecreateEventListener;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.reactive.engine.ReactiveExecutable;
+import org.hibernate.reactive.persister.collection.impl.ReactiveCollectionPersister;
+
+public class ReactiveCollectionRecreateAction extends CollectionAction implements ReactiveExecutable {
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java:25]: It seems to me that this should extend `CollectionRecreateAction`, to be consistent with what we do for entity-related actions, for example:

    public class ReactiveEntityDeleteAction extends EntityDeleteAction implements ReactiveExecutable

And that way we wouldn't need the copy/pasted `preRecreate()`/`postRecreate()` methods here.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRemoveAction.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+import org.hibernate.event.spi.PreCollectionRecreateEventListener;
+import org.hibernate.event.spi.PreCollectionRemoveEvent;
+import org.hibernate.event.spi.PreCollectionRemoveEventListener;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.reactive.engine.ReactiveExecutable;
+import org.hibernate.reactive.persister.collection.impl.ReactiveCollectionPersister;
+import org.hibernate.reactive.util.impl.CompletionStages;
+import org.hibernate.stat.spi.StatisticsImplementor;
+
+public class ReactiveCollectionRemoveAction extends CollectionAction implements ReactiveExecutable { ;
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRemoveAction.java:32]: Likewise, this would extend `CollectionRemoveAction`. (Unless there's something I'm missing.)

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+									loc = writeIndex( st, collection.getIndex( entry, i, this ), loc, session );
+								}
+								loc = writeElement( st, collection.getElement( entry ), loc, session );
+
+								collection.afterRowInsert( this, entry, i );
+							}
+					);
+					loop = loop.thenCompose( total -> reactiveConnection
+							.update( getSQLInsertRowString(), paramValues )
+							.thenApply( insertCount -> total + insertCount ) );
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:127]: This should also be a trampoline, I guess.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+								else {
+									writeElementToWhere( st, entry, loc, session );
+								}
+							}
+						}
+				);
+				loop = loop.thenCompose( total -> reactiveConnection
+						.update( getSQLDeleteRowString(), paramValues )
+						.thenApply( deleteCount -> total + deleteCount ) );
+			}
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:201]: Trampoline.

**Review by gavinking — CHANGES_REQUESTED** [review]
Excellent, this all looks very good to me, except I would like to see

1. use of trampolining via `CompletionStages.loop()`
2. inheritance of `Action`s from the non-reactive versions in order to eliminate copy/paste.


Thanks!!

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java (hunk 5 above)
Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:127]: See [CompletionStages#loop](https://github.com/hibernate/hibernate-reactive/blob/4844db0393e3738a5185da0ab3c709049a27137d/hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java#L155)

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+							.thenApply( insertCount -> total + insertCount ) );
+				}
+			}
+		}
+
+		return loop.thenAccept( total -> LOG.debugf( "Done inserting collection: %s rows inserted", total ) );
+	}
+
+	public CompletionStage<Integer> removeReactive(Serializable id, SharedSessionContractImplementor session)
+			throws HibernateException {
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:136]: Shouldn't we be consistent with ORM and return `CompletionStage<Void>` instead?
See [AbstractCollectionPersister#remove](https://github.com/hibernate/hibernate-orm/blob/20f6d4f5dbccd352d5a3e219121136cc53506f89/hibernate-core/src/main/java/org/hibernate/persister/collection/AbstractCollectionPersister.java#L1206)

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+	}
+
+	@Override
+	public CompletionStage<Void> reactiveUpdateRows(
+			PersistentCollection collection,
+			Serializable id,
+			SharedSessionContractImplementor session) throws HibernateException {
+
+		// Currently there is no "updateRows()" method in AbstractCollectionPersister
+		// Not sure if this method, reactiveUpdateRows() is needed.
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:214]: This comment is wrong: [BasicCollectionPersister#updateRows](https://github.com/hibernate/hibernate-orm/blob/5.4/hibernate-core/src/main/java/org/hibernate/persister/collection/AbstractCollectionPersister.java#L1746)

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/EagerElementCollectionForEmbeddableEntityTypeListTest.java [hunk 9]
```diff
@@ -0,0 +1,815 @@
...
+	@Embeddable
+	public static class Phone {
+
+		private String number;
+
+		private String country;
+
+		// FIXME: add nested @Embeddable
+		// @ElementCollection(fetch = FetchType.EAGER)
+		// private List<AlternativePhone> alternatives = new ArrayList<>();
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/EagerElementCollectionForEmbeddableEntityTypeListTest.java:768]: Can you remove this comments from the version we are going to merge?

Comment (DavideD) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/EagerElementCollectionForEmbeddableEntityTypeListTest.java:768]: This is probably going to be a different test class anyway

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+		if ( !isInverse && collection.isRowUpdatePossible() ) {
+
+			LOG.debugf( "Updating rows of collection: %s#%s", getNavigableRole().getFullPath(), id );
+
+			// update all the modified entries
+			// NOTE this method call uses a JDBC connection and will fail for Map ElementCollection type
+			// May also have to override getSQLDeleteString() to correctly parameterize to handle the SQL for the PG use case
+			// SEE getSQLInsertRowString() in this class for an example
+			int count = doUpdateRows( id, collection, session );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:230]: This is not the correct implementation anyway so it's probably going to fail if somebody reaches this code. I think we should throw an `UnsupportedOperationException` instead like we do in reactiveDeleteRows.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java [hunk 11]
```diff
@@ -55,4 +70,85 @@ protected ReactiveCollectionInitializer createSubselectInitializer(SubselectFetc
...
+			PersistentCollection collection, Serializable id, SharedSessionContractImplementor session) {
+		return CompletionStages.voidFuture();
+	}
+
+	private ReactiveConnection getReactiveConnection(SharedSessionContractImplementor session) {
+		return ( (ReactiveConnectionSupplier) session ).getReactiveConnection();
+	}
+
+	@Override
+	public CompletionStage<Integer> removeReactive(Serializable id, SharedSessionContractImplementor session)
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:85]: It should return `CompletionStage<Void>` like in ORM

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java [hunk 12]
```diff
@@ -55,4 +70,85 @@ protected ReactiveCollectionInitializer createSubselectInitializer(SubselectFetc
...
+			Expectation expectation = Expectations.appropriateExpectation( getDeleteAllCheckStyle() );
+			boolean callable = isDeleteAllCallable();
+
+			List<Object> params = new ArrayList<>();
+			params.add( id );
+			String sql = getSQLDeleteString();
+
+			return reactiveConnection.update( sql, params.toArray( new Object[0] ) );
+		}
+		return CompletionStages.completedFuture( -1 );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:109]: Even if we decide to return a `CompletionStage<Integer>` in this case, ORM returns 0 when it doesn't update any row. Not sure we should return -1 here.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java [hunk 13]
```diff
@@ -55,4 +70,85 @@ protected ReactiveCollectionInitializer createSubselectInitializer(SubselectFetc
...
+			return CompletionStages.voidFuture();
+		}
+
+		if ( !isRowDeleteEnabled() ) {
+			return CompletionStages.voidFuture();
+		}
+
+		// FIXME: Not sure what to implement here for OneToMany use-case
+		// Maybe nothing due to the ReactiveCollectionUpdate/Remove/Recreate Action classes perform necessary operations
+		// on the collection (basically delete and replace rather than operating on individual rows
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:141]: I think you can remove these comments, You had to add these methods because they are in the interface and I suspect, I'm not sure, we are going to work on this when adding support for `@ManyToMany` or tests associations with maps

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:141]: I think this would be enough:
```
// TODO: See AbstractCollectionPersister#insertRows
```

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Similar to {@link org.hibernate.persister.collection.AbstractCollectionPersister} in ORM
+ * <p>
+ *     The reactive classes already extend something so we cannot have an abstract class like in ORM.
+ * </p>
+ */
+public interface ReactiveCollectionPersister {
+
+	/**
+	 * Reactive version of {@link org.hibernate.persister.collection.AbstractCollectionPersister}.recreate()
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:23]: Is this a common way to link to the javadoc?
I usually do:
```
{@link org.hibernate.persister.collection.AbstractCollectionPersister#recreate}
```
Same for the rest of the javadoc.

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:23]: We could also remove the parameters from the javadoc since they don't have any helpful comment (I don't think they need one).

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:23]: This also applies to the javadoc of the other methods.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+	 */
+	CompletionStage<Void> recreateReactive(PersistentCollection collection, Serializable id, SharedSessionContractImplementor session);
+
+	/**
+	 * Reactive version of {@link org.hibernate.persister.collection.AbstractCollectionPersister}.remove()
+	 * @param id
+	 * @param session
+	 * @return Integer completion stage
+	 */
+	CompletionStage<Integer> removeReactive(Serializable id, SharedSessionContractImplementor session);
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:37]: This is the only method to return `CompletionStage<Integer>` instead of `CompletionStage<Void>` and it's also inconsistent with ORM.

**Review by DavideD — CHANGES_REQUESTED** [review]

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/collection/impl/ReactiveBatchingCollectionInitializerBuilder.java (hunk 2 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/collection/impl/ReactiveBatchingCollectionInitializerBuilder.java:27]: Just to fix the deprecation of getSettings(). I can change it back.

Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/collection/impl/ReactiveBatchingCollectionInitializerBuilder.java:27]: Oh I see. OK, that's fine.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java (hunk 3 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java:25]: This will require a change to the ORM actions which are currently marked as ```final```.  I'll test locally and create a PR that'll have to be back-ported to 5.4 at some point.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java (hunk 13 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:141]: got it

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java (hunk 14 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:23]: The best I can tell is:   ```{@link org.hibernate.persister.collection.AbstractCollectionPersister#insertRows()}```

thx

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:23]: As long as it is clickable

Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:23]: I can't find a way to create a clickable link to a method.

Best I could do is: ```   * Reactive version of {@link org.hibernate.persister.collection.AbstractCollectionPersister} recreate() method```

Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:23]: links to methods need to list the parameter types.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRemoveAction.java [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+//
+//	@Override
+//	public BeforeTransactionCompletionProcess getBeforeTransactionCompletionProcess() {
+//		return null;
+//	}
+//
+//	@Override
+//	public void afterDeserialize(SharedSessionContractImplementor session) {
+//
+//	}
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRemoveAction.java:121]: yes

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionUpdateAction.java [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+//
+//	@Override
+//	public BeforeTransactionCompletionProcess getBeforeTransactionCompletionProcess() {
+//		return null;
+//	}
+//
+//	@Override
+//	public void afterDeserialize(SharedSessionContractImplementor session) {
+//
+//	}
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionUpdateAction.java:152]: yes

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/EagerElementCollectionForEmbeddableEntityTypeListTest.java (hunk 9 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/EagerElementCollectionForEmbeddableEntityTypeListTest.java:768]: yes.. removed.  Already started on setting up the different test class.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 18]
```diff
@@ -260,44 +211,96 @@ protected String getSQLDeleteString() {
...
-							}
-							locOffset = writeElement( st, collection.getElement( entry ), locOffset, session );
+	/*
+	 * generates and returns parameter values for a reactiveRecreate() and reactiveInsertRows() methods
+	 * compatible with getSQLInsertRowString() and query string
+	 */
+	private Object[] paramValues(Object entry, PersistentCollection collection, Serializable id, SharedSessionContractImplementor session) {
+		return PreparedStatementAdaptor.bind(
+				st -> {
+					int i = 0;
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:266]: Something is wrong here.
In the ORM code the index `i` was incremented in the loop: https://github.com/hibernate/hibernate-orm/blob/cfc7b972501bf26713098abbbc36265a9b1ed46d/hibernate-core/src/main/java/org/hibernate/persister/collection/AbstractCollectionPersister.java#L1614

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 19]
```diff
@@ -260,44 +211,96 @@ protected String getSQLDeleteString() {
...
+					int loc = writeKey( st, id, 1, session );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:267]: In ORM, offset is initialized like this:
```
offset += expectation.prepare( st );
```
Shouldn't we do the same or something similar?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java [hunk 20]
```diff
@@ -106,7 +106,7 @@ private ReactiveConnection getReactiveConnection(SharedSessionContractImplemento
 
 			return reactiveConnection.update( sql, params.toArray( new Object[0] ) );
 		}
-		return CompletionStages.completedFuture( -1 );
+		return CompletionStages.completedFuture( 0  );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:109]: Shouldn't this return `CompletionStage<Void>`?

**Review by DavideD — COMMENTED** [review]
I have some questions about these new changes.
Not the end of the world but the commit message is  a bit misleading, there are other changes in this commit: it's not only about the trampoline.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java (hunk 20 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:109]: only reason is the ```reactiveConnection.update( sql, params.toArray( new Object[0] ) );``` returns ```CompletionStage<Integer>```

Guess we could convert to Void?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java (hunk 15 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:31]: k, I'll work through the changes

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java (hunk 20 above)
Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveOneToManyPersister.java:109]: OK, but all the other methods are Void, why this should be different?
I think I mentioned this inconsistency before. The method in `ReactiveCollectionPersister` should return `CompletionStage<Void>`: https://github.com/hibernate/hibernate-reactive/blob/384eb90a3a7e24f6e2b14e87d8d98689e4137075/hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java#L31

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java (hunk 15 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java]: fixed and pushed

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java (hunk 18 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:268]: looking

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+
+		CompletionStage<Integer> loop = CompletionStages.zeroFuture();
+		ReactiveConnection reactiveConnection = getReactiveConnection( session );
+
+		return CompletionStages.total(
+				entryArray(collection),
+				entry -> loop.thenCompose( total -> reactiveConnection
+						.update( getSQLInsertRowString(), paramValues( entry, collection, id ,session ) )
+						.thenApply( insertCount -> total + insertCount ) ) )
+				.thenAccept( total -> LOG.debugf( "Done inserting rows: %s inserted", total ) );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:225]: I haven't tested it, but I think this should be:
```
		return CompletionStages
				.total( entryArray( collection ), 
						entry -> reactiveConnection.update( getSQLInsertRowString(), paramValues( entry, collection, id,session ) ) )
				.thenAccept( total -> LOG.debugf( "Done inserting rows: %s inserted", total ) );
```

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:225]: By the way, if you create a `total` method in `CompletionStages` that accepts an Iterator or Iterable you can remove the conversion you need to do in `entryArray`

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+		ReactiveConnection reactiveConnection = getReactiveConnection( session );
+
+		boolean deleteByIndex = hasIndex && !indexContainsFormula;
+
+		return CompletionStages.total(
+				entryArray( collection, deleteByIndex ),
+				entry -> loop.thenCompose( total -> reactiveConnection
+						.update( getSQLDeleteRowString(), paramValues( entry, id ,session, deleteByIndex ) )
+						.thenApply( deleteCount -> total + deleteCount ) ) )
+				.thenAccept( total -> LOG.debugf( "Done removing rows: %s removed", total ) );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:168]: return CompletionStages
				.total( entryArray( collection, deleteByIndex ),
						entry -> reactiveConnection.update( getSQLDeleteRowString(), paramValues( entry, id ,session, deleteByIndex ) ) )
				.thenAccept( total -> LOG.debugf( "Done removing rows: %s removed", total ) );

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:168]: You can remove entryArray if you add a new `total` method in `CompletionStages` accepting an iterable

**Review by DavideD — CHANGES_REQUESTED** [review]

**Comment by blafond** [comment]
Closing.. replaced by https://github.com/hibernate/hibernate-reactive/pull/549
